curp/src/sync_manager.rs

```diff
@@ -144,137 +139,233 @@ impl<C: Command + 'static> SyncManager<C> {
+        #[allow(clippy::shadow_unrelated)] // clippy false positive
+        let (term, leader_id, prev_log_index, prev_log_term, entries, leader_commit) = state
+            .map_read(|state| {
+                let next_index = state.next_index[&connect.addr];
+                (
+                    state.term,
+                    0, // TODO: add leader_id
+                    next_index - 1,
+                    state.log[next_index - 1].term(),
+                    state.log[next_index..].to_vec(),
+                    state.commit_index,
                 )
-                .await
             });
+        let last_sent_index = prev_log_index + entries.len();
+        let req = match AppendEntriesRequest::new(
+            term,
+            leader_id,
+            prev_log_index,
+            prev_log_term,
+            entries,
+            leader_commit,
+        ) {
+            Err(e) => {
+                error!("unable to serialize append entries request: {}", e);
+                return;
+            }
+            Ok(req) => req,
+        };
+
+        // send append_entries request and receive response
+        debug!("append_entries sent to {}", connect.addr);
+        let resp = connect
+            .append_entries(req, Self::APPEND_ENTRIES_TIMEOUT)
+            .await;
+
+        #[allow(clippy::unwrap_used)]
+        // indexing of `next_index` or `match_index` won't panic because we created an entry when initializing the server state
+        match resp {
+            Err(e) => warn!("Heartbeat error: {}", e),
+            Ok(resp) => {
+                let resp = resp.into_inner();
+                let mut state = state.write();
+                if resp.success {
+                    // update match_index and next_index
+                    *state.match_index.get_mut(&connect.addr).unwrap() = last_sent_index;
+                    *state.next_index.get_mut(&connect.addr).unwrap() = last_sent_index + 1;
+
+                    let majority_cnt = (state.others.len() + 1) / 2 + 1;
+                    // update commit_index if needed
+                    if state.commit_index < last_sent_index
+                        // If the majority of servers has replicated the log, commit
+                        // + 1 means including the leader itself
+                        && 1 + state
+                            .others
+                            .iter()
+                            .filter(|addr| state.match_index[*addr] >= last_sent_index)
+                            .count()
+                            >= majority_cnt
+                    {
+                        debug!("commit_index updated to {last_sent_index}");
+                        state.commit_index = last_sent_index;
+                        if let Err(e) = commit_trigger.send(()) {
+                            error!("ae_trigger closed: {}", e);
+                        }
+                        if let Err(e) = ae_trigger.send(()) {
```

Comment:
When we commit a entry, why do we also trigger a "append_entry" request?

Reply:
It is an optimization to reduce commit latency. For example, if the leader sends heartbeats every 100ms, a log is received and replicated on the majority of the servers on the 20th ms. We want to trigger an "append_entries" request at once so that the commit index is carried to the followers immediately instead of waiting for another 80ms.

Reply:
It's a kind of "optimization" to make commit happens earlier on followers, but I'm not sure what's the benefit of the shorter commit latency.

Reply:
Removed.